File: 1000/Baekjoon_No1012.py
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]


def bfs(x, y):
    queue = [[x, y]]
    while queue:
        a, b = queue[0][0], queue[0][1]
        del queue[0]
        for i in range(4):
            move_x = a + dx[i]
            move_y = b + dy[i]
            if 0 <= move_x < n and 0 <= move_y < m and data[move_x][move_y] == 1:
                data[move_x][move_y] = 0
                queue.append([move_x, move_y])


for _ in range(int(input())):
    m, n, k = map(int, input().split())
    data = [[0]*m for _ in range(n)]
    count = 0

    for _ in range(k):
        x, y = map(int, input().split())
        data[y][x] = 1

    for i in range(n):
        for j in range(m):
            if data[i][j] == 1:
                bfs(i, j)
                data[i][j] = 0
                count += 1
    print(count)

"""
BFS를 이용한 문제
주어진 맵에서 0,0 부터 끝 까지 탐색하면서 1을 발견하면
해당 좌표를 0으로 만들고 count +1, 그 후 상하좌우로 이동가능한지 여부 확인하고
가능하다면 또 이동하면서 반복하기.
"""

File: 1000/Baekjoon_No1012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class b1012 {

    static int N, M, K;
    static int[] dx = {0, 0, 1, -1};
    static int[] dy = {1, -1, 0, 0};
    static int[][] board;
    static boolean[][] visited;
    static int count;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());

        while(T --> 0) {

            StringTokenizer st = new StringTokenizer(br.readLine());
            M = Integer.parseInt(st.nextToken());
            N = Integer.parseInt(st.nextToken());
            K = Integer.parseInt(st.nextToken());

            board = new int[M][N];
            visited = new boolean[M][N];
            count = 0;

            for (int i = 0; i < K; i++) {
                st = new StringTokenizer(br.readLine());
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());

                board[x][y] = 1;
            }

            for (int i = 0; i < M; i++) {
                for (int j = 0; j < N; j++) {
                    if (board[i][j] == 1 && !visited[i][j]) {
                        bfs(i, j);
                        count++;
                    }
                }
            }

            System.out.println(count);
        }
    }

    static void bfs(int x, int y) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[] {x, y});

        while(!queue.isEmpty()) {
           x = queue.peek()[0];
           y = queue.peek()[1];
           visited[x][y] = true;
           queue.poll();

           for(int i = 0; i < 4; i++) {
               int nx = x + dx[i];
               int ny = y + dy[i];

               if (nx >= 0 && ny >= 0 && nx < M && ny < N) {
                   if (!visited[nx][ny] && board[nx][ny] == 1) {
                       queue.add(new int[] {nx, ny});
                       visited[nx][ny] = true;
                   }
               }
           }
        }

    }
}
